#473 Add feature to load server environment variables from special file

MartinPeverelli opened the pull request:
Proposal for https://github.com/laravel/valet/issues/473

Lloople commented:
Hi @adamwathan @mattstauffer, is there any reason for not merge this pull request? I found that I need to set `env` variables per site and can't find a proper way. Adding this `.env.valet` file into the project looks pretty neat.

Thank you

mattstauffer commented:
@Lloople I don't see anything particularly wrong with the PR. What I did was close obvious "no" PRs, shut down all driver requests for now, and leave PR's like these open for Adam and Taylor to merge when they have some time to look over them and decide whether they fit within the spirit/intent of Valet. I think this being open is not a "no" but just a "taylor and adam haven't had time to dig into it yet".

iboldurev commented:
Please add this implementing

osrecio commented:
Hi, I think is a good approach for some kind of projects to have env variables and load it inside a project.

bramus commented:
Nice work! Would love to see this one get merged.

mattstauffer commented on cli/drivers/BasicValetDriver.php:
```diff
@@ -44,10 +44,12 @@ public function isStaticFile($sitePath, $siteName, $uri)
      */
     public function frontControllerPath($sitePath, $siteName, $uri)
     {
+        $this->loadServerEnvironmentVariables($sitePath, $siteName);
```
Since this seems to be happening in every `frontControllerPath` method, is it possible to instead just run this in `server.php` right before it calls `frontControllerPath`? (I don't know the code super well so there may be a good reason why not...)

drbyte replied:
Technically it probably could.

https://github.com/laravel/valet/blob/0323607850ec051ad4746409ed0992c5f24c37d9/server.php#L134-L147

MartinPeverelli commented:
@mattstauffer @drbyte I took both your advices and refactored this idea.
The file is now `.valet-env.php`, which makes it obvious that it should contain a PHP array (instead of the key=value pairs of regular dotfiles).
And I changed the method to be `public` and called it from `server.php` as requested, which allowed me to revert all the changes to all the Drivers.

Let me know what you think!

If this gets merged into master, I would also love to help with the documentation of this feature, how can I do that?

Lloople commented on cli/drivers/ValetDriver.php:
```diff
@@ -177,4 +177,28 @@ protected function isActualFile($path)
+        }
+
+        foreach ($variables[$siteName] as $key => $value) {
+            $_SERVER[$key] = $value;
```
Why not using `putenv($key.'='.$value)` here along with the already existing `$_SERVER`? 🤔

MartinPeverelli replied:
I was only trying to mimic Apache's way of loading env vars on the vhosts files, and if I'm not mistaken, that only sets values on`$_SERVER`.
I don't know if duplicating the values on both `$_SERVER` and `$_ENV` is useful or desirable, but as you say, we can add that single line to the loop if we do.

What say the rest of us?

drbyte commented:
@MartinPeverelli have you considered a way to allow for wildcard domain-names with this? 

I have a use-case where simply dropping in a non-domain-specific `.valet-env.php` file would make environment overrides easier. But I don't want to have to edit the file to set domain name/s inside it.
It's admittedly a little different use-case than your original need which was to do domain-specific custom overrides.

drbyte commented:
I've used this proposed PR on a project. It worked well.

I do think altering it to allow for being domain-agnostic would be nice too.

drbyte commented:
Waiting for @MartinPeverelli to merge [my PR to his valet fork](https://github.com/MartinPeverelli/valet/pull/1/files), which will update this PR, and then this could be merged.

MartinPeverelli commented:
@drbyte merged your PR into my fork! Sorry for the delay, hectic RL.

drbyte commented:
Thanks @MartinPeverelli 

I've been using this code for the past 3 weeks, and haven't run into any problems. It's definitely been helpful to keep project-specific environment variable configs in a separate `.valet-env.php` file in my app, instead of having to touch the valet-generated nginx configs.

@mattstauffer IMO this is safe to merge.

mattstauffer commented:
I think I got mixed up and thought this was adding `.env.valet` to each application, not `env-valet.php` to the root. Is there a reason we went this way? I was about to release but this feels weird.

MartinPeverelli commented:
The change to the file name was because the contents are a PHP array, instead of the key=value pair on the other .env files.

mattstauffer commented:
@MartinPeverelli sure... and you had to do PHP so you can differentiate them per domain. I remember now. Thanks!

drbyte commented:
Note: The filename is `.valet-env.php`
